shared_model/interfaces/common_objects/ordering.hpp

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_SHARED_MODEL_ORDERING_HPP
+#define IROHA_SHARED_MODEL_ORDERING_HPP
+
+#include "interfaces/common_objects/types.hpp"
+
+namespace shared_model {
+  namespace interface {
+
+    /// Provides ordering data for pagination
+    class Ordering final {
+    public:
+      Ordering& operator=(Ordering const&) = delete;
+      Ordering& operator=(Ordering&&) = delete;
+
+      Ordering();
+      Ordering(Ordering&&);
+      Ordering(Ordering const&);
+
+      enum struct Field : size_t{
+        created_time = 0,
+        position,
+        //--------------
+        max_value_count,
+      };
+
+      enum struct Direction : size_t{
+        ascending = 0,
+        descending,
+        //--------------
+        max_value_count,
+      };
+
+      using PriorityType = size_t;
+
+      struct OrderingEntry {
+        PriorityType priority;
+        Field field;
+        Direction direction;
+      };
+
+      /// keeps field and direction entry uniquely with a highest priority.
+      void insert(Field field, Direction direction, PriorityType priority);
+
+      /// drops all saved data.
+      void reset();
+
+      /// returns data sorted with priority from highest to lowest. Priority 0 ignored.
+      void get(OrderingEntry const*& orderingEntry, size_t& count) const;
+
+    private:
+      mutable OrderingEntry ordering_[(size_t)Field::max_value_count];
+      size_t count_;
+
+      inline void copy_(Ordering const& src) __attribute__((always_inline));
```

Comment:
`__attribute__` seems not a standard feature, but a GNU extension. I'm not sure if we can use it. some day we refused `@pragma once` as header guard because it is not part of the standard.